### DISCUSSION TITLE
How do you receive data from scripts/direct in the RemoteControl API?

### DISCUSSION THREAD

**Discussion by DrRichardDaystrom** [Stellarium/stellarium#3915] [Q&A] [answered]
Greetings, again!  I would like to call core.getJD() using scripts/direct in the RemoteControl API but I don't understand how to receive the resulting output.

In Python, this looks like ...

```
import requests

params_direct = {'code': 'core.getJD()'}
response = requests.post(url_stellarium + 'scripts/direct', data = params_direct)
```

Where does the output get returned?  I can't find it anywhere in the response structure (e.g., response.content, response.text, response.raw, etc.).  The status_code is 200 so the request appears to be successful.

Am I not calling core.getJD() correctly?  I've tried 'return core.getJD()' as the code for scripts/direct but that doesn't seem to work either.

Any insight would be appreciated.  I would like to use the functions in the Calendars plugin but I am stuck on how to receive their output.

**Comment by 10110111** [comment]
I don't think you can get back the return values from the code executed this way. To get current JD you can query `api/main/status`, taking `jday` item from `time` section in the output.

**Reply by DrRichardDaystrom** [reply]
Unfortunately, what I really want to do is use `Calendars.getJD()` to convert a (known) Gregorian date to its corresponding (unknown) Julian day so that I can POST that to `api/main/time/`.  However, to do so, I need to be able to receive the output from the first to pass as input to the second.

I was hoping to use Stellarium for this, since it seems to have all the required functionality (esp. for more esoteric calendars), rather than having to use separate packages like juliandate.

I appreciate your help - on this and my previous issue.  If anyone else knows how to do this, I am all ears.

**Reply by 10110111** [reply, marked as answer]
Why not just pass the value you got to `core.setJDay()` directly in the same JS code?

**Reply by DrRichardDaystrom** [reply]
Good idea.  That works.  As a proof of concept, here is a snippet that gets the JDay and then increments the JDay by one ...

```
params_direct = {'code': 'JD = core.getJDay(); core.setJDay(JD+1);'}
response = requests.post(url_stellarium + 'scripts/direct', data = params_direct)
```

I am trying to do as much as possible in Python since I am working on some Jupyter notebooks for students and I want them to be able to inspect all the variables in that environment.  However, your initial suggestion of using `api/main/status` to retrieve the JDay after it has been set in this fashion will work for my purposes.

I'm still working on understanding the calling syntax for `Calendars.getJD()` but now I can poke at it to see if I can figure it out.

I also have question about how `api/main/time` interprets the JDay when the timezone is set to LTST but I will bang my head on that a little more before opening up a new question.

Thanks again for all your help.  I am very much enjoying playing with the program.  Keep up the good work.

**Reply by 10110111** [reply]
> how `api/main/time` interprets the JDay when the timezone is set to LTST

Julian day is timezone-agnostic, see its definition.

**Reply by DrRichardDaystrom** [reply]
Yes, understood.  I guess my issue is that I have to account for the time offset due to longitude when calling `api/main/time`.

When I call `api/main/time` with a round JD, like 2110695, it is only solar noon when the longitude is zero.

If I want it to be solar noon at my observation location, I have to add (or subtract) the offset due to the local longitude.

It makes sense that it works that way - there's no such thing as a "local" JD.  I'm working on accounting for this right now.

**Reply by DrRichardDaystrom** [reply]
Ugh, it is harder than that.  Round JDs correspond to 12:00:00 UTC - not local solar noon at the prime meridian.

Therefore, if I want to set the time to local solar noon (at the prime meridian or elsewhere), I first need to figure out the UTC time of local solar noon and use that to modify the JD passed to `api/main/time` (as well as accounting for the time offset due to longitude).

I presume I can use the equation of time to figure this out but I was hoping that I could just use LTST to handle this (and do the longitude correction myself).  I know see why that was naive.

If there is an easy way to do this, please let me know.

**Reply by DrRichardDaystrom** [reply]
Hmm, this appears to be a dead end.  I can't find any way to access the EquationOfTime plugin programmatically - either from JS or the Remote Control API.

I guess my next idea is to use `api/stelaction/do?id='actionNext_Transit'` for the Sun to determine the UTC time of solar noon.

**Reply by gzotti** [reply]
You can also add an issue request for us to make core->getSolutionEquationOfTime() scriptable in the main object. Nobody has asked for that so far. Or you make this change and send a PR.

**Reply by DrRichardDaystrom** [reply]
Thanks @10110111  & @gzotti .  I (almost) figured out my problem - specifically, how to set the time to local true solar noon.

Instead of using the Equation of Time plugin, which is not scriptable, I used `objects/info` in the RemoteControl plugin API to get the transit time for the Sun.  The only other thing to be careful of is to fetch the gmtShift from `main/status` - esp. for dates before 1847 for which the timezone is automatically set to LMST.

Here is my (almost) solution, in Python.  It seems to find solar noon correctly, even for really ancient dates, but the LTST time is +2s(-5d) in the year 1000, +20s(+3d) in the year 0, and +48s(+10d) in the year -1000.  I presume that the change in dates is due to calendar reforms but I am not sure the cause of the residual differences from 12:00:00 LTST.  Any insights would be welcome.

```
import requests
import juliandate as jd

# specify URL to Stellarium API
url_stellarium = 'http://localhost:8090/api/'

# specify observation location ... move your choice to the end of the list
obs_loc = {'latitude':'+51.47783','longitude':'-0.00139','altitude':'68',
           'name':'Greenwich','country':'England','planet':'Earth'}
obs_loc = {'latitude':'+33.3564','longitude':'-116.865','altitude':'1712',
           'name':'Palomar','country':'USA','planet':'Earth'}

# specify observation date as a Julian day (jd) starting at midday ... move your choice to the end of the list
obs_jd = jd.from_gregorian(2024, 6, 20) + 0.5  # after the introduction of timezones
obs_jd = jd.from_gregorian(1800, 6, 21) + 0.5  # before the introduction of timezones

# set Stellarium location to obs_loc
response_set_loc = requests.post(url_stellarium + 'location/setlocationfields', data = obs_loc)

# set Stellarium time to obs_jd & timerate to zero (stopped)
params_time = {'time': str(obs_jd), 'timerate': '0'}
response_time = requests.post(url_stellarium + 'main/time', data = params_time)

# 'focus' on Sun
response_focus = requests.post(url_stellarium + 'main/focus', data = {'target':'Sun','mode':'mark'})

# get object info for Sun
response_obj_info = requests.get(url_stellarium + 'objects/info', data = {'name':'Sun','format':'json'})

# get main status to see what timezone Stellarium is using
response_status = requests.get(url_stellarium + 'main/status')

# set time to transit ... transit-dhr includes timezone and equation of time ... must account for both
transit_jd = obs_jd + (response_obj_info.json()['transit-dhr']-12)/24 - response_status.json()['time']['gmtShift']
params_transit = {'time': str(transit_jd), 'timerate': '0'}
response_transit = requests.post(url_stellarium + 'main/time', data = params_transit)

# set Stellarium time zone to LTST to verify that the true solar time is 12:00:00
params_tz = {'id':'StelCore.currentTimeZone', 'value':'LTST'}
response_tz = requests.post(url_stellarium + 'stelproperty/set', data = params_tz)
```
